typescript/sdk/src/core/app.ts

```diff
@@ -1,74 +1,56 @@
 import { Inbox, Outbox } from '@abacus-network/core';
 
 import { AbacusApp } from '../app';
-import { MultiProvider } from '../provider';
+import { AbacusAddresses } from '../contracts';
 import { ChainMap, ChainName, Remotes } from '../types';
-import { objMap } from '../utils';
 
-import {
-  CoreContractAddresses,
-  CoreContractSchema,
-  CoreContracts,
-} from './contracts';
+import { CoreContracts, coreFactories } from './contracts';
 import { environments } from './environments';
 
-export const CoreEnvironments = Object.keys(environments);
 export type CoreEnvironment = keyof typeof environments;
 export type CoreEnvironmentChain<E extends CoreEnvironment> = Extract<
   keyof typeof environments[E],
   ChainName
 >;
 
+export type CoreContractsMap<Chain extends ChainName> = {
+  [local in Chain]: CoreContracts<Chain, local>;
+};
+
 export class AbacusCore<Chain extends ChainName = ChainName> extends AbacusApp<
-  CoreContracts<Chain>,
+  CoreContracts<Chain, Chain>,
   Chain
 > {
-  constructor(
-    addresses: {
-      [local in Chain]: CoreContractAddresses<Chain, local>;
-    },
-    multiProvider: MultiProvider<Chain>,
-  ) {
-    super(CoreContracts, addresses, multiProvider);
+  constructor(contractsMap: CoreContractsMap<Chain>) {
+    super(contractsMap);
   }
 
-  static fromEnvironment<E extends CoreEnvironment>(
-    name: E,
-    multiProvider: MultiProvider<any>, // TODO: fix networks
+  static fromEnvironment<Env extends CoreEnvironment>(
+    env: Env,
   ): AbacusCore<any> {
-    return new AbacusCore(environments[name], multiProvider);
-  }
-
-  extendWithConnectionManagers<T>(
-    config: ChainMap<Chain, T>,
-  ): ChainMap<Chain, T & { abacusConnectionManager: string }> {
-    return objMap(config, (chain, config) => ({
-      ...config,
-      abacusConnectionManager:
-        this.getContracts(chain).abacusConnectionManager.address,
-    }));
-  }
-
-  // override type to be derived from chain key
-  getContracts<Local extends Chain>(
-    chain: Local,
-  ): CoreContractSchema<Chain, Local> {
-    return super.getContracts(chain) as any;
+    const addressesMap = environments[env] as ChainMap<
+      CoreEnvironmentChain<Env>,
+      AbacusAddresses
+    >;
+    const contractsMap: CoreContractsMap<any> = this.buildContracts(
+      addressesMap,
+      coreFactories,
+    );
+    return new AbacusCore(contractsMap);
   }
 
   // override type to be derived from chain key
-  getAddresses<Local extends Chain>(
-    chain: Local,
-  ): CoreContractAddresses<Chain, Local> {
-    return super.getAddresses(chain) as any;
+  getContracts<Local extends Chain>(chain: Local): CoreContracts<Chain, Local> {
```

Comment:
Is this still neccessary?

Reply:
yes, because of the `App<AbacusApp<CoreContracts<Chain, Chain>,  ...`